Pull the trace struct out of the low-level interface.

vext01 opened the pull request:
This is a companion change to https://github.com/softdevteam/yktrace/pull/1

ltratt commented:
It's not clear from the code: is this merging 2 MIR blocks into 1 MIR blocks or 2 LLVM blocks into 1 LLVM block?

vext01 commented:
Upon re-visiting this, I think I've slightly misunderstood. It's not merging anything at all.

It changes this:
```
    /---- B_2
B_1
    \---- B_3
```

To:

```
B_1 -> B_2
```

If it knows that the assertion can never cause `B_3` to be reachable. In this case, `B_3` is not code-genned at all.

I think we do still want to disable this, so that the structure of the binary is the same as the MIR, but the wording of the comment and commit message need to be revised.

(sorry).

ltratt commented:
I don't think we need to disable this: all it means is that a MIR block *x* won't have a label *x* in the binary. But that's equivalent to having a label *x* in the binary and never PTing it.

vext01 commented:
I think I've just come to the same conclusion myself...

Shall we move on *without* it and maybe revisit it if there's evidence that this is indeed a problem for us? In my eyes, the smaller the divergence from upstream the better!

ltratt commented:
Yes, I'd prefer to have the fewest changes we can get away with. So let's close this?

vext01 commented:
(sorry for the noise)

vext01 committed:
Pull the trace struct out of the low-level interface.

By pulling the struct from here, and moving it into yktrace, we can
offer the user a more abstracted interface, including a Drop and an
Iterator.

Also revise some comments while we are here.

ltratt commented on src/test/run-pass/yk_swt/collect-traces-threaded.rs:
```diff
@@ -25,29 +25,32 @@ fn main() {
-    for i in 0..t.len() {
-        let loc = t.loc(i);
+    for i in 0..len {
+        if i > isize::max_value() as usize {
```
Wouldn't it be easier to hoist this outside the loop? Something like `assert!(len < (isize::max_value() as usize))`.

vext01 replied:
Indeed!

ltratt commented:
Which PR should be done first?

vext01 commented:
> Which PR should be done first?

I think we should decide if we like the API outlined in https://github.com/softdevteam/yktrace/pull/1, and if we do, then we merge this ykrustc PR first, then the yktrace PR. This is because the yktrace PR depends on the compiler changes.

BTW, you may have noticed the FIXME. I'd like to restructure our libcore code at some point, but I didn't want to mix it in with this PR and make it confusing to review. I hope that's OK.

vext01 commented:
Your comment was addressed in my recent push BTW, so this is also ready for re-review.

ltratt commented:
The updates look good, but I guess we'll need to wait for the `Option` change to `start_tracing` to be fixed here too?

vext01 commented:
Correct

ltratt commented:
Do we need to do something to this PR before considering it for merging?

vext01 commented:
If you have no further comments, we can squash this.

ltratt commented:
Please squash.

vext01 commented:
squashed and revised commit message.

ltratt commented:
bors r+

vext01 commented:
```
+ ./x.py clean
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
```

vext01 commented:
bors retry

vext01 commented:
OK it's building on worker1 (it failed on worker4). I restarted all of the workers just in case that's needed to pick up the "fresh clone" settings we changed yesterday.

I'll be keeping an eye on this, as I'm not totally sure if it's fixed it for all workers or not...
